### PR TITLE
Issue#16990: Cast -> AllToAll -> Cast fails with random output

### PR DESCRIPTION
**Pull request by shaahji**
Issue#16990: Cast -> AllToAll -> Cast fails with random output

### Description
The python test was implemented wrong for the boolean casting scenario.

Also, parameterized the test case instead of running multiple scenarios within the same test case.

### Motivation and Context
Fixes Issue #16990

### CODE REVIEW

File: onnxruntime/test/python/onnxruntime_test_collective.py [hunk 1]
```diff
@@ -154,42 +155,50 @@ def _create_alltoall_ut_model_for_boolean_tensor(
...
-            expected_output = np.zeros((128, 128), dtype=np_elem_type)
-            for _ in range(size - 1):
-                expected_output = np.concatenate((expected_output, np.ones((128, 128), dtype=np_elem_type) * (_ + 1)))
-            np.testing.assert_allclose(outputs[0], expected_output, err_msg="all gather on axis0: result mismatch")
+        input = np.ones((128, 128), dtype=np.float32) * rank
+        outputs = ort_sess.run(None, {"X": input})
+
+        expected_output = np.zeros((128, 128), dtype=np_elem_type)
+        for _ in range(size - 1):
+            expected_output = np.concatenate((expected_output, np.ones((128, 128), dtype=np_elem_type) * (_ + 1)))
```
Comment (wschin) [inline onnxruntime/test/python/onnxruntime_test_collective.py:195]: move this logic into
```
def all_gather_implementation(tensor_on_device_0, tensor_on_device_1, ...):
  ...
```

or use a more generic signature
```
def all_gather_implementation(*tensors_on_devices):
  ...
```

File: onnxruntime/test/python/onnxruntime_test_collective.py [hunk 2]
```diff
@@ -223,44 +232,50 @@ def test_all_gather_axis1(self):
...
 
     def test_all_to_all_bool(self):
         rank, _ = self._get_rank_size()
 
-        if rank == 0:
-            x = [True, True, True, True]
-        else:
-            x = [False, False, False, False]
-
+        x = [rank == 0] * 4
```
Comment (wschin) [inline onnxruntime/test/python/onnxruntime_test_collective.py:278]: Please revert this back. It's harder to read.

File: onnxruntime/test/python/onnxruntime_test_collective.py (hunk 1 above)
Comment (wschin) [inline onnxruntime/test/python/onnxruntime_test_collective.py:195]: I honestly don't understand this computation. Could you give an example as a comment in code? E.g.,
```
# Assume you have 2 devices.
# Device 0 has 2-D tensor x=[ [0, 1], [2, 3] ]
# Device 1 has 2-D tensor y=[ [4, 5], [6, 7] ]
# AllToAll(x, y) may produce
# [ [?, ?], [?, ?] ] on device 0 and
# [ [?, ?], [?, ?] ] on device 1.
```

File: onnxruntime/test/python/onnxruntime_test_collective.py [hunk 3]
```diff
@@ -223,44 +232,50 @@ def test_all_gather_axis1(self):
...
+        )
+        rank, size = self._get_rank_size()
+        ort_sess = ort.InferenceSession(
+            model.SerializeToString(),
+            providers=["CUDAExecutionProvider", "CPUExecutionProvider"],
+            provider_options=[{"device_id": str(rank)}, {}],
+        )
 
+        if communication_elem_type == TensorProto.BOOL:
+            input = np.ones((128, 128), dtype=np_elem_type) if rank == 0 else np.zeros((128, 128), dtype=np_elem_type)
```
Comment (wschin) [inline onnxruntime/test/python/onnxruntime_test_collective.py:254]: Please elaborate this computation with example. Thanks.

File: onnxruntime/test/python/onnxruntime_test_collective.py [hunk 4]
```diff
@@ -223,44 +232,50 @@ def test_all_gather_axis1(self):
...
+        else:
             input = np.ones((128, 128), dtype=np_elem_type) * rank
-            outputs = ort_sess.run(None, {"X": input})
 
-            expected_output = np.zeros((int(128 / size), 128), dtype=np_elem_type)
+            expected_output = np.zeros((128 // size, 128), dtype=np_elem_type)
             for _ in range(size - 1):
                 expected_output = np.concatenate(
-                    (expected_output, np.ones((int(128 / size), 128), dtype=np_elem_type) * (_ + 1))
+                    (expected_output, np.ones((128 // size, 128), dtype=np_elem_type) * (_ + 1))
```
Comment (wschin) [inline onnxruntime/test/python/onnxruntime_test_collective.py:268]: Replace this with a formal python implementation of AllToAll. The  computation here is hard to understand. Maybe also refactorize the test following this type:
```python
tensor_on_device_0 = ...
tensor_on_deivce_1 = ...
...

if rank == 0:
  tensor_to_fed_to_ort = tensor_on_device_0
elif rank == 1:
  tensor_to_fed_to_ort = tensor_on_device_1
...

out = ort_session.run(tensor_to_fed_to_ort)

baseline_out = all_to_all_implementation(tensor_on_device_0, tensor_on_device_1, tensor_on_device_2, ...)

assert_all_close(baseline_out[rank], out)
```

File: onnxruntime/test/python/onnxruntime_test_collective.py (hunk 1 above)
Comment (shaahji) [reply onnxruntime/test/python/onnxruntime_test_collective.py:195]: Splitting this logic will make reading the test implementation harder. Moreover, the logic is only building data for a single rank so a function that takes data for all ranks as input doesn't make sense.

Adding documentation on input/output per rank to elaborate on the logic.

```
        # Input for size == 2
        # rank = 0:
        #   Input:  [[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]]
        #   Output: [[0, 0, 0, 0], [0, 0, 0, 0], [1, 1, 1, 1], [1, 1, 1, 1]]
        #
        # rank = 1:
        #   Input:  [[1, 1, 1, 1], [1, 1, 1, 1], [1, 1, 1, 1], [1, 1, 1, 1]]
        #   Output: [[0, 0, 0, 0], [0, 0, 0, 0], [1, 1, 1, 1], [1, 1, 1, 1]]
```

File: onnxruntime/test/python/onnxruntime_test_collective.py (hunk 2 above)
Comment (shaahji) [reply onnxruntime/test/python/onnxruntime_test_collective.py:278]: It's Pythonic. Reverted anyways! :)

File: onnxruntime/test/python/onnxruntime_test_collective.py (hunk 1 above)
Comment (shaahji) [reply onnxruntime/test/python/onnxruntime_test_collective.py:195]: You are commenting in all gather implementation but your comment reads AllToAll. Little confused about where this comment was intended for AllGather or AllToAll?

File: onnxruntime/test/python/onnxruntime_test_collective.py (hunk 3 above)
Comment (shaahji) [reply onnxruntime/test/python/onnxruntime_test_collective.py:254]: Done!

File: onnxruntime/test/python/onnxruntime_test_collective.py (hunk 4 above)
Comment (shaahji) [reply onnxruntime/test/python/onnxruntime_test_collective.py:268]: Your approach would create tensors for each device adding to the confusion. I have added explicit comments to state what the inputs are and the expected output and why.

File: onnxruntime/test/python/onnxruntime_test_collective.py (hunk 1 above)
Comment (shaahji) [reply onnxruntime/test/python/onnxruntime_test_collective.py:195]: Added comments on inputs/outputs for all operations.

**Comment by shaahji** [comment]
@wschin Can we get this to finish line?

**Review by wschin — DISMISSED** [review]
Thanks a lot for improving the documentation part. Writing more documents might decrease the development of specific PRs, but it will make all downstream works and collaborations easier. Thanks again. :)

**Comment by shaahji** [comment]
@wschin @souptc PR ready for review

P.S. CI failures are unrelated to this change.

**Review by souptc — APPROVED** [review]

**Pull request merged**